drivers/gpu/drm/rp1/rp1-vec/rp1_vec.c

```diff
@@ -158,79 +214,156 @@ static void rp1vec_connector_destroy(struct drm_connector *connector)
 	drm_connector_cleanup(connector);
 }
 
-static const struct drm_display_mode rp1vec_modes[4] = {
+/*
+ * Check the mode roughly matches something we can generate.
+ * The choice of hardware TV mode depends on total lines and frame rate.
+ * Within each hardware mode, allow pixel clock, image size and offsets
+ * to vary, up to a maximum horizontal active period and line count.
+ * Don't check sync timings here: the HW driver will sanitize them.
+ */
+
+static enum drm_mode_status rp1vec_mode_valid(struct drm_device *dev,
+					      const struct drm_display_mode *mode)
+{
+	int prog	  = !(mode->flags & DRM_MODE_FLAG_INTERLACE);
+	int fake_31khz	  = prog && mode->vtotal >= 500;
+	int vtotal_2fld	  = mode->vtotal << (prog && !fake_31khz);
+	int vdisplay_2fld = mode->vdisplay << (prog && !fake_31khz);
+	int real_clock	  = mode->clock >> fake_31khz;
+
+	/* Check pixel clock is in a sane range */
+	if (real_clock < 6016)
+		return MODE_CLOCK_LOW;
+	else if (real_clock > 27000)
+		return MODE_CLOCK_HIGH;
+
+	/* Try to match against the 525-line 60Hz mode (System M) */
+	if (vtotal_2fld >= 500 && vtotal_2fld < 544 && vdisplay_2fld <= 486 &&
```

Comment:
The line-count test here is awfully permissive, the reason was to allow command-line modes like `640x480M` (which is a CVT mode incompatible with TV systems). Now we have `cmode` that's not a good reason, so I think I shall make it stricter again.